modules/300-prometheus/hooks/generate_istio_mtls_cert.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2022 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/deckhouse/deckhouse/go_lib/certificate"
+)
+
+const (
+	mTLSCrtPath = "prometheus.internal.prometheusScraperIstioMTLS.certificate"
+	mTLSKeyPath = "prometheus.internal.prometheusScraperIstioMTLS.key"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	Queue: "/modules/prometheus/generate_istio_mtls_cert",
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       "istio_secret_ca",
+			ApiVersion: "v1",
+			Kind:       "Secret",
+			FilterFunc: applySecertIstioCAFilter,
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-istio"},
+				},
+			},
+			NameSelector: &types.NameSelector{
+				MatchNames: []string{"cacerts"},
+			},
+		},
+		{
+			Name:       "secret_istio_mtls",
+			ApiVersion: "v1",
+			Kind:       "Secret",
+			FilterFunc: applySecertIstioMTLSFilter,
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-monitoring"},
+				},
+			},
+			NameSelector: &types.NameSelector{
+				MatchNames: []string{"prometheus-scraper-istio-mtls"},
+			},
+		},
+	},
+	Schedule: []go_hook.ScheduleConfig{
+		{Name: "cron", Crontab: "0 0 * * *"}, // every day at 00:00 am
+	},
+}, generateMTLSCertHook)
+
+func applySecertIstioMTLSFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	secret := &v1.Secret{}
+	err := sdk.FromUnstructured(obj, secret)
+
+	if err != nil {
+		return nil, fmt.Errorf("can't convert ca secret to secret struct: %v", err)
+	}
+
+	var cert certificate.Certificate
+	var certBytes, keyBytes []byte
+	var ok bool
+	if certBytes, ok = secret.Data["tls.crt"]; ok {
+		cert.Cert = string(certBytes)
+	} else {
+		return nil, fmt.Errorf("can't get certificate from secert %v", secret.Name)
+	}
+
+	if keyBytes, ok = secret.Data["tls.key"]; ok {
+		cert.Key = string(keyBytes)
+	} else {
+		return nil, fmt.Errorf("can't get key from secert %v", secret.Name)
+	}
+	return cert, nil
+}
+
+func applySecertIstioCAFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
```

Comment:
As I see it, the code is the same for both filters. Only certificate and key keys are different. 

Let's rewrite the code to reuse the generic logic using decorators.
```go
func filterCertificate(key, cert string, obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
  ...
}

func filterCa(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
  return filterCertificate("ca-cert.pem", "ca-key.pem", obj)
}

func filterCert(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
  return filterCertificate("tls.crt", "tls.key", obj)
}
```

Reply:
They also return different types.
	var cert certificate.Certificate
	var cert certificate.Authority
it is a problem for decorator.

Reply:
Yeah, overlooked it. Thanks.